smarter default init

ohayon opened the pull request:
Fixes https://github.com/fastlane/fastlane/issues/815

ohayon committed:
New Default Init Stuff

KrauseFx commented on Gemfile:
```diff
@@ -3,19 +3,18 @@ source 'https://rubygems.org'
-  gem "supply", path: "../supply"
-  gem "scan", path: "../scan"
-end
+gem "fastlane_core", path: "../fastlane_core"
```
Did you mean to commit the `cd ..`?

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -57,13 +56,30 @@ def copy_existing_files
+    end
 
+    def ask_for_apple_id
+      ask('Your Apple ID (fastlane@krausefx.com): '.yellow)
```
Should prefix the email with `(e.g. fastlane@...`

ohwutup replied:
sorry, not so sure what you mean here. do you mind explaining a bit more? :smile:

KrauseFx replied:
Add the `e.g.` in front of the email :wink:

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +88,27 @@ def generate_appfile
+      Produce.config = FastlaneCore::Configuration.create(Produce::Options.available_options, produce_options_hash)
+      begin
+        ENV['PRODUCE_APPLE_ID'] = Produce::Manager.start_producing
+      rescue Exception => exception
```
Instead of using `Exception`, just use

```ruby
rescue => exception
```

This way you don't catch keyboard interrupts and similar exceptions.

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +88,27 @@ def generate_appfile
+        ENV['PRODUCE_APPLE_ID'] = Produce::Manager.start_producing
+      rescue Exception => exception
+        Helper.log.info 'It looks like that App Name has already been taken, please enter an alternative.'.yellow
+        Produce.config[:app_name] = ask("App Name:".yellow)
```
Missing space after `:`

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +88,27 @@ def generate_appfile
+      begin
+        ENV['PRODUCE_APPLE_ID'] = Produce::Manager.start_producing
+      rescue Exception => exception
+        Helper.log.info 'It looks like that App Name has already been taken, please enter an alternative.'.yellow
```
There can all kinds of exception, the name is just one of them. Easiest way is to add
```ruby
if ex.to_s.include?("bla bla")
```

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +88,27 @@ def generate_appfile
+        Helper.log.info 'It looks like that App Name has already been taken, please enter an alternative.'.yellow
+        Produce.config[:app_name] = ask("App Name:".yellow)
+        Produce.config[:skip_devcenter] = true
+        ENV['PRODUCE_APPLE_ID'] = Produce::Manager.start_producing
```
What if the name is taken again?

ohwutup replied:
thats a good question... do we want to allow them to continue to enter names until they get something that works? or should we crash after a couple of tries?

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
       end
 
-      @tools[:sigh] = true if agree("Do you want to use 'sigh', which will maintain and download the provisioning profile for your app? (y/n)".yellow, true)
+      if agree("Do you want to use 'sigh', which will maintain and download the provisioning profile for your app? (y/n)".yellow, true)
```
Let's get rid of that, as we now have `match`

ohwutup replied:
get rid of sigh all together? i thought that we had talked about not forcing people into using match? is it possible with match to store the profiles locally? i think we should talk about how the code signing should be set up by default. i don't know if we can expect everyone to have an available private repo somewhere.

KrauseFx replied:
Mh, good point. Let's just recommend that you use `match`. Let's leave `sigh` on there then.

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
+
+    def default_enable_other_tools
+      enable_deliver
+      # enable_snapshot
```
hu?

ohwutup replied:
?

KrauseFx replied:
Why is `enable_snaphshot` commented out?

ohwutup replied:
i had it commented just because i wanted to chat with you about what default behavior should be fore this.

i know we had talked about whether or not we should implement this by default but i don't remember what the conclusion was that we came to

KrauseFx replied:
We don't enable it by default, but have it prepared in the `Fastfile`, so the user can easily enable it

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
+    end
+
+    def default_enable_other_tools
+      enable_deliver
```
We're already calling that from somewhere else, right?

ohwutup replied:
not sure i understand? i think that you commented on the same file a couple times maybe? :smile:

KrauseFx replied:
No, you're calling it twice

ohwutup replied:
oh, i think it was being called from the `ask_to_enable_other_tools` method which i just got rid of because it wasn't getting called anymore

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
+    end
+
+    def enable_sigh
+      @tools[:sigh] = true
```
We don't need this method

ohwutup replied:
why not?

KrauseFx replied:
Because we always use `sigh`, so we don't need to set the `@tools` hash, right?

ohwutup replied:
I put this in here because I wanted to be able to call `enable_sigh` from the `default_enable_other_tools` method while the same line of code was being called in `ask_to_enable_other_tools` (which i just removed as it's not being used anymore). maybe there is something that i'm missing about this `@tools` hash.

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
+      require 'snapshot/setup'
+      Snapshot::Setup.create(folder)
+
+      @tools[:snapshot] = true
```
We don't enable `snapshot` by default, instead we tell the user to run `snapshot init` once to set it up

ohwutup replied:
this is in an `enable_snapshot` method that gets called from the existing method `ask_to_enable_other_tools` i just abstracted it out of that method so that i could call it from another place also.

KrauseFx replied:
:+1:

ohwutup replied:
clear!

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
-    def generate_fastfile
-      template = File.read("#{Helper.gem_path('fastlane')}/lib/assets/FastfileTemplate")
+    def generate_fastfile(scheme: nil)
+      template = File.read("#{Helper.gem_path('fastlane')}/lib/assets/DefaultFastfileTemplate")
```
What happened to the old file? Is it still there? We don't need it I guess?

ohwutup replied:
still there for now, i was very careful to leave everything as it was and only change things. being that this is my first bit set of changes to the project i wanted to just go back after and clean stuff like that up! :smile_cat:

KrauseFx replied:
We don't need the old file anymore :wink:

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -87,36 +124,63 @@ def ask_to_enable_other_tools
+      Helper.log.info "Loading up 'deliver', this might take a few seconds"
+      require 'deliver'
+      require 'deliver/setup'
+      options = FastlaneCore::Configuration.create(Deliver::Options.available_options, {})
```
I haven't tested it, but this will take the apple id, app identifier and the team id from the Appfile, right?

ohwutup replied:
right now its not doing team as far as i know, but i haven't tested with an apple id that has access to multiple teams. do you have access to an account like that?

ohwutup replied:
adding team name into Appfile happens on the first run not on configuration

KrauseFx commented on lib/assets/DefaultFastfileTemplate:
```diff
@@ -24,32 +24,11 @@ default_platform :ios
-
-  desc "Submit a new Beta Build to Apple TestFlight"
-  desc "This will also make sure the profile is up to date"
-  lane :beta do
```
Why don't we have a `beta` lane any more?

ohwutup replied:
Per some conversation above, I had removed the original FastfileTemplate and replaced it with this DefaultFastfileTemplate. It looks like i deleted the lanes here, but based on your suggestion, I had deleted the file, so git is noticing it as a rename.

KrauseFx commented on lib/assets/DefaultFastfileTemplate:
```diff
@@ -24,32 +24,11 @@ default_platform :ios
-
-  desc "Runs all the tests"
-  lane :test do
-    snapshot
```
Why don't have a `test` lane any more?

ohwutup replied:
same thing as the above comment

KrauseFx commented on lib/fastlane/setup/setup.rb:
```diff
@@ -22,7 +22,7 @@ def run
       else
         raise "Couldn't find platform '#{platform}'"
       end
-      ask_for_crash_reporting
```
:no_entry_sign:

ohwutup replied:
?

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +79,29 @@ def generate_appfile
     end
 
+    def default_run_produce
+      Helper.log.info "Running produce..."
```
This will always run, right? This must be super scary for users, they run `fastlane` the first time and it's already **modifying** and **creating** things on the Developer Portal. 

Do we already have the summary table which the user can confirm? We should be very transparent of what `fastlane` will do next before it actually does it.

ohwutup replied:
yeah, i agree this is pretty weird. i think that i thought we had talked about this in one of our initial meetings about this and thought it should happen, i am happy to get rid of this part of the feature.

ohwutup replied:
I just removed the call to this method, but i think it might be nice to have a really easy `produce` command for later on during development when someone just wants to run `produce` and not answer questions. what do you think? should i move this code somewhere  into produce to have a less-interactive version of the action?

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +79,29 @@ def generate_appfile
+      project = FastlaneCore::Project.new(config)
+
+      produce_options_hash = {
+          app_name: project.default_app_name
```
Wrong spacing here

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -72,6 +79,29 @@ def generate_appfile
+        ENV['PRODUCE_APPLE_ID'] = Produce::Manager.start_producing
+      rescue => exception
+        if exception.to_s.include?("The App Name you entered has already been used")
+          Helper.log.info 'It looks like that App Name has already been taken, please enter an alternative.'.yellow
```
`enter an alternative`, the user never entered one, so this message should be more clear, something like
> It looks like that App Name of your project has already been taken on iTunes Connect or the Apple Developer Portal. Please provide a name for your app:

KrauseFx commented on Gemfile:
```diff
@@ -3,7 +3,7 @@ source 'https://rubygems.org'
 gemspec
 
-if `cd ..; git remote -v`.include?("countdown")
+# if `cd ..; git remote -v`.include?("countdown")
```
Does this change belong here?

KrauseFx commented on lib/fastlane/setup/setup.rb:
```diff
@@ -22,7 +22,7 @@ def run
         raise "Couldn't find platform '#{platform}'"
       end
-      ask_for_crash_reporting
+      FastlaneCore::CrashReporting.enable
```
Let's disable it for now!

KrauseFx commented on lib/fastlane/setup/setup_ios.rb:
```diff
@@ -34,6 +33,26 @@ def run
+      config = {}
+      FastlaneCore::Project.detect_projects(config)
+      @project = FastlaneCore::Project.new(config)
+      @project.default_app_identifier
```
Is this necessary?

KrauseFx committed:
Improved table of summary

ohayon committed:
refactor create appfile method and fix fastfiletemplate

ohayon committed:
disable crash reporting by default

ohayon committed:
comment on project ivar access

KrauseFx committed:
Automatic app detection on Portal / iTC

KrauseFx committed:
Merge branch 'master' of https://github.com/krausefx/fastlane into feature/produce

KrauseFx committed:
Merge branch 'feature/produce' into default-init

ohayon committed:
change some naming

ohayon committed:
revert gemfile changes

KrauseFx committed:
Added automatic creating of new apps and storing the team ID in the Appfile

KrauseFx committed:
Merge branch 'master' into default-init

# Conflicts:
#	lib/fastlane/setup/setup_ios.rb

KrauseFx committed:
Improved Fastfile template

KrauseFx committed:
Improved iOS init

Refactored code
Removed all the @ and switched to attr_accessor
Removed useless code (sigh, xctool)
Restored cocoapods and carthage functionality
Updated all the setup tests

KrauseFx committed:
Caching of project name and bundle id

KrauseFx committed:
Fixed tests

KrauseFx committed:
Moving of metadata too

KrauseFx committed:
Updated fastlane_core dependency

KrauseFx commented:
:rocket:

KrauseFx committed:
Fixed test
